### DISCUSSION TITLE
Custom variables scoping in components

### DISCUSSION THREAD

**Discussion by nicriz** [mccode-dev/McCode#1689] [Q&A] [answered]
Hi all,

I am working on a custom version of Guide_gravity.comp that keeps track of the times t at which there is a reflection and the neutron perpendicular energy at the same point. In order to store these values I added    
```
    double reflection_times[15]; //record up to 15 reflection times
    double transverse_E[15]; //record up to 15 reflection energies
```
to the `typedef struct Gravity_guide_Vars`. Then I basically followed the example of `N_reflection` to initialize my variables and, finally, to assign the values after a reflection:
```
        int current_index;
        current_index = (int)GVars.N_reflection[0];
        GVars.reflection_times[current_index]=t;
        GVars.transverse_E[current_index]= 5.227e-6*q/(2*V2Q); //0.5*m*Vz^2 [meV]
        GVars.N_reflection[0]++;
```
Here the full component:
[Guide_gravity_multiple_collisions.comp.txt](https://github.com/user-attachments/files/16942287/Guide_gravity_multiple_collisions.comp.txt)

Now, I have minimum working instrument with two guides:
[test_amplitude.instr.txt](https://github.com/user-attachments/files/16942027/test_amplitude.instr.txt)
Here an example of output:
```
Guide 1
N Reflections: 1
Time 0: 0.00356201 Energy 0: 0.000257431 
Time 1: 0 Energy 1: 0 
Time 2: 0 Energy 2: 0 
Time 3: 0 Energy 3: 0 
Time 4: 0 Energy 4: 0 
Guide 2
N Reflections: 3
Time 0: 0.0222356 Energy 0: 0.000132608 
Time 1: 0.0357154 Energy 1: 0.000114036 
Time 2: 0.0460601 Energy 2: 0.000132608 
Time 3: 0 Energy 3: 0 
Time 4: 0 Energy 4: 0 
GTime 0: 0.00356201 GEnergy 0: 0.000257431
GTime 1: 0.0222356 GEnergy 1: 0.000132608
GTime 2: 0.0357154 GEnergy 2: 0.000114036
GTime 3: 0.0460601 GEnergy 3: 0.000132608
```
But the goal is to implement this in a larger model with 33 divergent guides:
[NNBAR_new.instr.txt](https://github.com/user-attachments/files/16942062/NNBAR_new.instr.txt)
The problem is that the latter throws an error at compilation, despite both are using the exact same component
```
./NNBAR_new.c: In function 'class_Guide_gravity_multiple_collisions_trace':
./NNBAR_new.c:19719:26: error: 'Gravity_guide_Vars_type' {aka 'struct Gravity_guide_Vars'} has no member named 'reflection_times'
19719 |     for (i=0; i<15; GVars.reflection_times[i++] = 0);
      |                          ^
./NNBAR_new.c:19720:26: error: 'Gravity_guide_Vars_type' {aka 'struct Gravity_guide_Vars'} has no member named 'transverse_E'
19720 |     for (i=0; i<15; GVars.transverse_E[i++] = 0);
      |                          ^
./NNBAR_new.c:19839:14: error: 'Gravity_guide_Vars_type' {aka 'struct Gravity_guide_Vars'} has no member named 'reflection_times'
19839 |         GVars.reflection_times[current_index]=t;
      |              ^
./NNBAR_new.c:19840:14: error: 'Gravity_guide_Vars_type' {aka 'struct Gravity_guide_Vars'} has no member named 'transverse_E'
19840 |         GVars.transverse_E[current_index]= 5.227e-6*q*q/(4*V2Q*V2Q); //0.5*m*Vz^2 [meV]
      |              ^
./NNBAR_new.c:19890:73: error: 'Gravity_guide_Vars_type' {aka 'struct Gravity_guide_Vars'} has no member named 'reflection_times'
19890 |                               printf("Time 0: %g Energy 0: %g \n", GVars.reflection_times[0], GVars.transverse_E[0]);
...
```
So my question is, **what in the instrument file could influence the scoping of component variables ?**. The implementation seems identical to the testing instrument.

I run the instruments as: 
`mcrun -c -I /mnt/nfs/home/rizzi_n/NNBAR_old/NNBARnew/run_folder -n 100 test_amplitude.instr -g`

McStas Version: 3.4.55 (clean from conda channel)
Linux version 4.18.0-553.5.1.el8_10.x86_64

**Comment by nicriz** [comment, marked as answer]
For anyone who is facing the same problem: the fact that I was loading a standard Guide_gravity before my custom component was messing up the struct definition. Converting them to the custom guide fixed it for me.

**Comment by willend** [comment]
@nicriz thanks for reporting that - so the solution was to use another name / set of defines for your modified struct?

**Comment by willend** [comment]
@nicriz I will by the way reopen the discussion (first of a kind in fact, contrats :-D ) to make it look a little more active here :-)

**Reply by nicriz** [reply]
Hello again,

Sorry @willend for not replying, I didn't see the notification.
I would like to resume this thread because I want to solve this issue, instead of just avoiding the problem :)

So, last time the problem was that I didn't change the tydef struct Gravity_guide_Vars_type, so when i was loading both Guide_gravity AND Guide_gravity_multiple_collisions, the compile was confused by the same struct name with different definitions. 

Now I have fixed that, giving different names to the typedef struct and updating all the others structures, but I keep getting an error, which I do not understand. Where is the clash in the definitions?

```
CFLAGS=
INFO: Recompiling: ./NNBAR_ILL.out
./NNBAR_ILL.c:13127:3: error: unknown type name 'Guide_gravity_multiple_collisions_Vars_type'
13127 |   Guide_gravity_multiple_collisions_Vars_type  GVarsGlobal;
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./NNBAR_ILL.c: In function 'class_Guide_gravity_multiple_collisions_init':
./NNBAR_ILL.c:14348:21: error: request for member 'compcurname' in something not a structure or union
14348 |   strcpy(GVarsGlobal.compcurname, NAME_CURRENT_COMP);
      |                     ^

./NNBAR_ILL.c:14352:5: warning: implicit declaration of function 'Guide_gravity_multiple_collisions_Init'; did you mean 'class_Guide_gravity_multiple_collisions_init'? [-Wimplicit-function-declaration]
14352 |     Guide_gravity_multiple_collisions_Init(&GVarsGlobal,
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |     class_Guide_gravity_multiple_collisions_init
./NNBAR_ILL.c:14358:39: error: request for member 'A' in something not a structure or union
14358 |     if (!G) for (i=0; i<5; GVarsGlobal.A[i++] = 0);
```

A simple example file:
[Guide_gravity_multiple_collisions.comp.txt](https://github.com/user-attachments/files/18523585/Guide_gravity_multiple_collisions.comp.txt)
[NNBAR_ILL.instr.txt](https://github.com/user-attachments/files/18523587/NNBAR_ILL.instr.txt)

**Reply by nicriz** [reply]
This was caused by the ifndef at the beginning of the component, which was the same for Gravity_guide :)

**Comment by willend** [comment]
Great, sorry for not responding earlier and thanks for figuring this out yourself @nicriz. ;-) Let me know if everything is aok now?

**Reply by nicriz** [reply]
Looks like it's the case! :)
